python/cuml/dask/linear_model/linear_regression.py

```diff
@@ -88,217 +81,105 @@ def _func_create_model(sessionId, **kwargs):
-            worker_info[wf[0]]["r"],
-            key="%s-%s" % (key, idx),
-            workers=[wf[0]]))
-            for idx, wf in enumerate(self.linear_models)])
+        because dask might have distributed the partitions of X and y
+        into different workers.
 
-        wait(list(linear_fit.values()))
-        raise_exception_from_futures(list(linear_fit.values()))
+        Parameters
+        ----------
+        X : dask cuDF dataframe (n_rows, n_features)
+            Features for regression
+        y : dask cuDF (n_rows, 1)
+            Labels (outcome values)
+        force_colocality: boolean (True: re-distributes the partitions
+                          of X and y to force the co-locality of
+                          the partitions)
+        """
 
-        comms.destroy()
+        # todo: add check for colocality in case force_colocality=False
 
-        self.local_model = self.linear_models[0][1].result()
-        self.coef_ = self.local_model.coef_
-        self.intercept_ = self.local_model.intercept_
+        X = self.client.persist(X)
+        y = self.client.persist(y)
 
-    def _fit_with_colocality(self, X, y):
-        input_futures = self.client.sync(extract_colocated_ddf_partitions,
-                                         X, y, self.client)
-        workers = list(input_futures.keys())
+        data = MGData.colocated(data=(X, y), client=self.client)
+        self.datatype = data.datatype
 
         comms = CommsContext(comms_p2p=False)
-        comms.init(workers=workers)
+        comms.init(workers=data.workers)
 
-        worker_info = comms.worker_info(comms.worker_addresses)
+        data.calculate_parts_to_sizes(comms)
+        self.ranks = data.ranks
 
         n_cols = X.shape[1]
-        n_rows = 0
 
-        self.rnks = dict()
-        partsToSizes = dict()
         key = uuid1()
-        for w, futures in input_futures.items():
-            self.rnks[w] = worker_info[w]["r"]
-            parts = [(self.client.submit(
-                LinearRegression._func_get_size_cl,
-                future,
-                workers=[w],
-                key="%s-%s" % (key, idx)).result())
-                for idx, future in enumerate(futures)]
-
-            partsToSizes[worker_info[w]["r"]] = parts
-            for p in parts:
-                n_rows = n_rows + p
-
-        key = uuid1()
-        self.linear_models = [(w, self.client.submit(
+        linear_models = [(w, self.client.submit(
             LinearRegression._func_create_model,
             comms.sessionId,
+            self.datatype,
             **self.kwargs,
             workers=[w],
             key="%s-%s" % (key, idx)))
-            for idx, w in enumerate(workers)]
+            for idx, w in enumerate(data.workers)]
 
         key = uuid1()
-        linear_fit = dict([(worker_info[wf[0]]["r"], self.client.submit(
-            LinearRegression._func_fit_colocated,
+        linear_fit = dict([(data.worker_info[wf[0]]["r"], self.client.submit(
```

Comment:
Still trying to brainstorm how we can wrap this common paradigm in the future... Maybe `dask_utils.map_function_over_worker_objects(func, linear_models, data)` takes a function to run on each model_object and automatically passes through parts_to_sizes, key, worker_info and other stuff we use every time? But that wouldn't help in the model creation case (submit calls above)

Reply:
This is actually mapping over the partitions that were extracted from the input, which are being grouped by worker.  The only reason the `worker_info` is relevant is because it needs the rank of the workers holding each list of partitions. It might be nice to have some function that can take a `dict({worker: [part]})` and create the parts to sizes, etc.. automatically (since the parts_to_sizes is not actually needed outside of the fit).